WalletWasabi.Tests/UnitTests/EventSourcing/InMemoryEventRepositoryTests.cs

```diff
@@ -0,0 +1,452 @@
+		}
+
+		[Fact]
+		public async Task AppendEventsAsync_InterleavingConflict_Async()
+		{
+			// Arrange
+			var events_a_0 = new[] { new WrappedEvent(1) };
+			var events_b_0 = new[] { new WrappedEvent(1) };
+			var events_a_1 = new[] { new WrappedEvent(2) };
+
+			// Act
+			async Task Action()
+			{
+				await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "a", events_a_0);
+				await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "b", events_b_0);
+				await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "a", events_a_1);
+				await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "a", events_a_1);
+			}
+
+			// Assert
+			await Assert.ThrowsAsync<OptimisticConcurrencyException>(Action);
+			Assert.True((await EventRepository.ListEventsAsync(nameof(TestRoundAggregate), "a"))
+				.SequenceEqual(events_a_0.Concat(events_a_1)));
+			Assert.True((await EventRepository.ListEventsAsync(nameof(TestRoundAggregate), "b"))
+				.SequenceEqual(events_b_0));
+			Assert.True((await EventRepository.ListAggregateIdsAsync(nameof(TestRoundAggregate)))
+				.SequenceEqual(new[] { "a", "b" }));
+		}
+
+		[Fact]
+		public async Task AppendEvents_AppendIsAtomic_Async()
+		{
+			// Arrange
+			var events_1 = new[] { new TestWrappedEvent(1, "a"), new TestWrappedEvent(2, "a") };
+			var events_2 = new[] { new TestWrappedEvent(2, "b"), new TestWrappedEvent(3, "b") };
+
+			// Act
+			async Task Action()
+			{
+				await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events_1!);
+				await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events_2!);
+			}
+
+			// Assert
+			await Assert.ThrowsAsync<OptimisticConcurrencyException>(Action);
+			Assert.True((await EventRepository.ListEventsAsync(nameof(TestRoundAggregate), "1"))
+				.Cast<TestWrappedEvent>().SequenceEqual(events_1));
+			Assert.True((await EventRepository.ListAggregateIdsAsync(nameof(TestRoundAggregate)))
+				.SequenceEqual(new[] { "1" }));
+		}
+
+		[Theory]
+		[InlineData(nameof(TestInMemoryEventRepository.AppendedSemaphore))]
+		public async Task AppendEvents_CriticalSectionConflicts_Async(string conflictAfter)
+		{
+			// Arrange
+			var events_1 = new[] { new TestWrappedEvent(1, "a"), new TestWrappedEvent(2, "a") };
+			var events_2 = new[] { new TestWrappedEvent(2, "b"), new TestWrappedEvent(3, "b") };
+
+			// Act
+			async Task Append1()
+			{
+				await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events_1!);
+			}
+			async Task Append2()
+			{
+				switch (conflictAfter)
+				{
+					case nameof(TestInMemoryEventRepository.AppendedSemaphore):
+						Assert.True(TestEventRepository.AppendedSemaphore.Wait(_semaphoreWaitTimeout));
+						break;
+
+					default:
+						throw new ApplicationException($"unexpected value conflictAfter: '{conflictAfter}'");
+				}
+				await EventRepository.AppendEventsAsync(nameof(TestRoundAggregate), "1", events_2!);
```

Comment:
callback system is absolutely critical to test correctly interaction between two threads in critical sections

Reply:
sorry I don't know what you mean, the line number changed, which method is that?

Reply:
Firstly I need to explain how multi-threaded debugging and testing works. In general if you go from one thread to two threads the complexity of testing grows quadratically. {number_of_commands} ^ {number_of_threads} because any thread can meet with any other thread at any command. 

I'm doing two threaded tests. The two threads in theory can meet at any command. I identified "critical sections" which are points of potential conflict of those two threads (the "empty helper methods"). I have written tests to make sure I tested all combination of how the two threads meet in those critical sections.